controller/program/src/instructions/swap_exact_in.rs

```diff
@@ -0,0 +1,402 @@
+    cpi::price::lp::cpi_price_exact_in, instructions::price::exact_in::PriceExactInIxArgs,
+};
+use inf1_svc_jiminy::cpi::{cpi_lst_to_sol, cpi_sol_to_lst};
+use jiminy_cpi::{
+    account::AccountHandle,
+    pda::{PdaSeed, PdaSigner},
+    program_error::{ProgramError, INVALID_ACCOUNT_DATA, NOT_ENOUGH_ACCOUNT_KEYS},
+};
+use sanctum_spl_token_jiminy::{
+    instructions::transfer::transfer_checked_ix_account_handle_perms,
+    sanctum_spl_token_core::{
+        instructions::transfer::{NewTransferCheckedIxAccsBuilder, TransferCheckedIxData},
+        state::{
+            account::{RawTokenAccount, TokenAccount},
+            mint::{Mint, RawMint},
+        },
+    },
+};
+
+use crate::{
+    instructions::sync_sol_value::sync_sol_val_with_retval,
+    pricing::NewPpIxPreAccsBuilder,
+    svc::NewSvcIxPreAccsBuilder,
+    verify::{verify_not_rebalancing_and_not_disabled, verify_pks},
+    Accounts, Cpi,
+};
+
+#[inline]
+pub fn process_swap_exact_in(
+    accounts: &mut Accounts<'_>,
+    args: &IxArgs,
+    cpi: &mut Cpi,
+) -> Result<(), ProgramError> {
+    if args.amount == 0 {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::ZeroValue).into());
+    }
+    let (ix_prefix, suf) = accounts
+        .as_slice()
+        .split_first_chunk()
+        .ok_or(NOT_ENOUGH_ACCOUNT_KEYS)?;
+
+    let ix_prefix = IxPreAccs(*ix_prefix);
+
+    let pool_state = *ix_prefix.pool_state();
+    let lst_state_list = *ix_prefix.lst_state_list();
+    let inp_lst_mint = *ix_prefix.inp_lst_mint();
+    let out_lst_mint = *ix_prefix.out_lst_mint();
+    let inp_pool_reserves = *ix_prefix.inp_pool_reserves();
+    let out_pool_reserves = *ix_prefix.out_pool_reserves();
+
+    let list = LstStatePackedList::of_acc_data(accounts.get(lst_state_list).data())
+        .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidLstStateListData))?;
+
+    let inp_lst_state = list
+        .0
+        .get(args.inp_lst_index as usize)
+        .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidLstIndex))?;
+
+    let inp_lst_state = unsafe { inp_lst_state.as_lst_state() };
+
+    let inp_lst_mint_acc = accounts.get(inp_lst_mint);
+    let inp_token_prog = inp_lst_mint_acc.owner();
+
+    let out_lst_state = list
+        .0
+        .get(args.out_lst_index as usize)
+        .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidLstIndex))?;
+
+    let out_lst_state = unsafe { out_lst_state.as_lst_state() };
+
+    let out_lst_mint_acc = accounts.get(out_lst_mint);
+    let out_token_prog = out_lst_mint_acc.owner();
+
+    let expected_inp_reserves = create_raw_pool_reserves_addr(
+        inp_token_prog,
+        &inp_lst_state.mint,
+        &inp_lst_state.pool_reserves_bump,
+    )
+    .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidReserves))?;
+
+    let expected_out_reserves = create_raw_pool_reserves_addr(
+        out_token_prog,
+        &out_lst_state.mint,
+        &out_lst_state.pool_reserves_bump,
+    )
+    .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidReserves))?;
```

Comment:
you can approx. halve the code here, since you're performing the same procedure on both `inp` and `out` lst, by doing an [array.map](https://doc.rust-lang.org/std/primitive.array.html#method.map)

Reply:
> like an array of tuple (lst_index, lst_mint: AccountHandle)

Yes. The types can be anything you like, up to you to decide which ones make the most sense. For the input array type, which params vary across the different array entries? For the output array type, what is the output you want from each one?

```rust
let [
    (inp_lst_state, inp_token_prog, expected_inp_reserves),
    (out_lst_state, out_token_prog, expected_out_reserves),
] = [
    (args.inp_lst_index as usize, inp_lst_mint),
    (args.out_lst_index as usize, out_lst_mint),
].map(|(idx, mint)| ...);
```